core/capabilities/vault/capability.go

```
@@ -171,6 +173,18 @@ func validateWriteRequest(id string, encryptedSecrets []*vaultcommon.EncryptedSe
 			return errors.New("secret must have encrypted value set at index " + strconv.Itoa(idx) + ":" + req.Id.String())
 		}
 
+		// Validate that the encrypted value can be decrypted with the Vault public key
+		cipherBytes, err := hex.DecodeString(req.EncryptedValue)
+		if err != nil {
+			return errors.New("failed to decode encrypted value at index " + strconv.Itoa(idx) + ":" + err.Error())
+		}
+		if publicKey != nil { // Public key can be nil if gateway cache isn't populated yet
```

sorry, and if the public key is nil - does this check still make sense? Or should we have stopped earlier?

---

publicKey can be nil on gateway nodes if they still haven't retrieved and cached the Vault Public key.
This validation function runs in 2 places, both as an additional redundant layer of fast error checking:
1. Gateway Nodes: Here, we only do validation with public key if public key was already cached on the node. Otherwise, just skip the validation
2. Vault Capability Nodes: These nodes always have vault public key readily available. So they will definitely do the validation.

As I said above, both the above checks are just for upfront error checking and fail fast.
The real validation of these values happens inside the OCR round of Vault DON.

The above validation "hex.DecodeString(req.EncryptedValue)" is useful even if publicKey is nil. It fails fast, if user provides a garbage value there which isn't even hex encoded.